### PR TITLE
Add statemint as trustable chain for pendulum + integration tests

### PR DESCRIPTION
**Pull request by RustNinja**
PR related to https://github.com/pendulum-chain/pendulum/issues/174 and https://github.com/pendulum-chain/pendulum/issues/191

- configured xcm that Statemint is trustable source for reserve transfer assets for Pendulum runtime
- add integration tests and basic configuration for network
    - parachain (pendulum-runtime)
    - polkadot (polkadot-runtime)
    - statemint (statemint-runtime)
- use [xcm-emulator](https://github.com/shaunxw/xcm-simulator/blob/master/xcm-emulator/Cargo.toml) instead of [xcm-simulator](https://github.com/paritytech/polkadot/blob/release-v0.9.37/xcm/xcm-simulator/src/lib.rs)
- add dependency of statemint-runtime from [parity cumulus](https://github.com/paritytech/cumulus).
configure statemint runtime for MockNet.
- register asset with id 1984 to simulate USDT as sufficient on system parachain.
- cover bidirectional transfer from Statemint to Pendulum and back.
- unit test to check transfer DOT from pendulum to relay chain

### CODE REVIEW

File: runtime/integration-tests/pendulum/src/tests.rs
Comment (ebma) [inline runtime/integration-tests/pendulum/src/tests.rs:19]: Same here, what is this fee value about? Does it have to be this value? Please add a comment above that explains its usage

File: runtime/integration-tests/pendulum/src/tests.rs
Comment (ebma) [inline runtime/integration-tests/pendulum/src/tests.rs:22]: Nit: `transfer_dot_from_relay_chain_to_pendulum()`

File: runtime/integration-tests/pendulum/src/tests.rs
Comment (ebma) [inline runtime/integration-tests/pendulum/src/tests.rs:37]: Does it need to be `into().into()` or is it a typo and one `into()` is enough?

Comment (ebma) [reply runtime/integration-tests/pendulum/src/tests.rs:37]: Also, I can see [here](https://github.com/paritytech/polkadot/blob/645723987cf9662244be8faf4e9b63e8b9a1b3a3/xcm/xcm-simulator/example/src/lib.rs#L210) in the example that they wrapped it with the `X1()` junction. Shouldn't we do this as well?

File: runtime/integration-tests/pendulum/src/tests.rs
Comment (ebma) [inline runtime/integration-tests/pendulum/src/tests.rs:70]: Nit: `transfer_dot_from_pendulum_to_relay_chain`

File: runtime/integration-tests/pendulum/src/tests.rs
Comment (ebma) [inline runtime/integration-tests/pendulum/src/tests.rs:77]: This should be called `before_bob_free_balance` probably.

Comment (ebma) [reply runtime/integration-tests/pendulum/src/tests.rs:77]: Maybe we can extract the `dot(100)` to an extra variable and put it in line 74. as
```
let expected_base_balance = dot(100)
```

and then you can also replace the `dot(100)` in the assert statement in l. 131 with that variable.

File: runtime/integration-tests/pendulum/src/tests.rs
Comment (ebma) [inline runtime/integration-tests/pendulum/src/tests.rs:18]: Maybe rename `TEN` to `TEN_UNITS`

Comment (ebma) [reply runtime/integration-tests/pendulum/src/tests.rs:18]: Or remove it and use `10 * UNIT` in the places it was used instead.

File: runtime/integration-tests/pendulum/src/lib.rs
Comment (ebma) [inline runtime/integration-tests/pendulum/src/lib.rs:2]: Why is the macro annotation just for `polkadot_test_net`?

File: runtime/integration-tests/pendulum/src/tests.rs
Comment (ebma) [inline runtime/integration-tests/pendulum/src/tests.rs:136]: It's a bit hard to understand what this test is about. Why is the asset `XCM(1)` 'incorrect' and what would we expect? Maybe add some comments to the code.

File: runtime/integration-tests/pendulum/src/tests.rs
Comment (ebma) [inline runtime/integration-tests/pendulum/src/tests.rs:307]: Why do we need to have this line? Maybe add a comment

File: runtime/integration-tests/pendulum/src/tests.rs
Comment (ebma) [inline runtime/integration-tests/pendulum/src/tests.rs:336]: Maybe move the imports to the top of the function

File: runtime/integration-tests/pendulum/src/polkadot_test_net.rs
Comment (ebma) [inline runtime/integration-tests/pendulum/src/polkadot_test_net.rs:1]: Let's maybe rename this file to `network_setup.rs` and combine it with the other `setup.rs` file so that we only have one. I think it's hard to find good naming because they are doing similar things, that's why it might make sense to combine them in the first place.

**Review by ebma — CHANGES_REQUESTED** [review]
Overall this looks good and has some very valuable changes. Great job @RustNinja 👍 

I think I would restructure the folder, so that it is `/runtime/integration-tests/src/pendulum` and not `/runtime/integration-tests/pendulum/src`. Maybe we don't need to have that `pendulum` subfolder anyways.

Also, @RustNinja I assume we would need to change the XCM config for amplitude as well to support non-native assets? Which XCM assets do we actually support with these changes, I suppose it's actually any asset from any chain that we have an open channel to?

**Comment by RustNinja** [comment]
> Overall this looks good and has some very valuable changes. Great job @RustNinja 👍
> 
> I think I would restructure the folder, so that it is `/runtime/integration-tests/src/pendulum` and not `/runtime/integration-tests/pendulum/src`. Maybe we don't need to have that `pendulum` subfolder anyways.
> 
> Also, @RustNinja I assume we would need to change the XCM config for amplitude as well to support non-native assets? Which XCM assets do we actually support with these changes, I suppose it's actually any asset from any chain that we have an open channel to?

Here is the all changes that i made in one [pr](https://github.com/pendulum-chain/pendulum/pull/198)

I would like to keep integration tests separate(in different packages) for each runtime(Pendulum/Amplitude/Foucoco).
The reason is that probably one day we will have a different version of polkadot for different networks. 0.39 for Pendulum and 0.43 for Foucoco. So it will be easy to support integration tests in different packages for each network.

> Also, @RustNinja I assume we would need to change the XCM config for amplitude as well to support non-native assets? Which XCM assets do we actually support with these changes, I suppose it's actually any asset from any chain that we have an open channel to?

@ebma 
by this changes in [xcm_config](https://github.com/pendulum-chain/pendulum/pull/192/files#diff-91848e42611fbf16b8a970242187d57e75083383e231b2a458f8404b2d08e920) Pendulum supports only `USDT` asset_id 1984 from Statemint. 
We can enable them everything but we need a task and approval from product to do it.

Amplitude does not have any [open channel](https://kusama.subscan.io/xcm_dashboard) with other chains. so we do not need to change xcm config for amplitude yet.

File: runtime/integration-tests/pendulum/src/tests.rs
Comment (ebma) [reply runtime/integration-tests/pendulum/src/tests.rs]: Nice, thanks for the great comment. 
Should we move it above the declaration of the test function ie line 135 maybe?

**Review by ebma — COMMENTED** [review]
@RustNinja so if I understand correctly, USDT (1984) is the only asset that will work because it's the only asset that we implemented a conversion for in the `xcm_config` [here](https://github.com/pendulum-chain/pendulum/blob/1dbf5375e961c0fd52eac67ed13ccccab61de821/runtime/pendulum/src/xcm_config.rs#L61-L70)?

File: runtime/integration-tests/pendulum/src/tests.rs
Comment (ebma) [reply runtime/integration-tests/pendulum/src/tests.rs]: Okay I see. If we don't need it, you can also remove it again. Or also add it to the [AccountId](https://github.com/paritytech/polkadot/blob/645723987cf9662244be8faf4e9b63e8b9a1b3a3/xcm/xcm-simulator/example/src/lib.rs#L211) to make it consistent 😅

File: runtime/integration-tests/pendulum/src/polkadot_test_net.rs
Comment (RustNinja) [reply runtime/integration-tests/pendulum/src/polkadot_test_net.rs:1]: @ebma 
I found that parallel finance integration tests more well organised
here is the link:
https://github.com/parallel-finance/parallel/tree/master/integration-tests/src

but pretty similar way Acala did also.
https://github.com/AcalaNetwork/Acala/blob/master/runtime/integration-tests/src/lib.rs

File: runtime/integration-tests/pendulum/src/polkadot_test_net.rs
Comment (RustNinja) [reply runtime/integration-tests/pendulum/src/polkadot_test_net.rs]: renamed to fn units(amount)

File: runtime/integration-tests/pendulum/src/tests.rs
Comment (RustNinja) [reply runtime/integration-tests/pendulum/src/tests.rs]: Done.

**Comment by RustNinja** [comment]
> here
@ebma 

Exactly. We declared [here](https://github.com/pendulum-chain/pendulum/blob/1dbf5375e961c0fd52eac67ed13ccccab61de821/runtime/pendulum/src/xcm_config.rs#L62-L70) that we trust(support) as reserve backed assets only two multilocation: 

`MultiLocation::parent()` for Relay-chain(our parrent)
and 
```
MultiLocation::new(
	1,
	X3(
		Parachain(statemint::PARA_ID),
		PalletInstance(statemint::ASSET_PALLET_ID),
		GeneralIndex(statemint::USDT_ASSET_ID),
	),
))
```
no other assets until we get approval from product team.

File: runtime/integration-tests/pendulum/src/tests.rs
Comment (RustNinja) [reply runtime/integration-tests/pendulum/src/tests.rs]: Done.

**Comment by ebma** [comment]
Now that you added the changes to the Foucoco XCM config as well, maybe we should go ahead and change it also for Amplitude? Would it look any different for Amplitude compared to the other two?

**Comment by RustNinja** [comment]
> Now that you added the changes to the Foucoco XCM config as well, maybe we should go ahead and change it also for Amplitude? Would it look any different for Amplitude compared to the other two?

@ebma 

I added the changes to Foucoco to first upgrade foucoco runtime before Pendulum runtime upgrade and check that it will works with DOT transfer from relay chain to our chain and back.
Anyway it works in integration tests and i just want to use this new xcm settings for Foucoco.
for Amplitude there is no any other reserve backed assets from other chain that we should support so i decide to not touch amplitude runtime without actual requirements for it. 
But you right. There should not be any difference in xcm config for pendulum and amplitude.
So let me know if you fell that i should also change amplitude in this PR.

File: runtime/pendulum/src/lib.rs
Comment (TorstenStueber) [inline runtime/pendulum/src/lib.rs:17]: This is not used.

**Review by TorstenStueber — CHANGES_REQUESTED** [review]

File: Cargo.toml
Comment (RustNinja) [reply Cargo.toml]: Thanks. Done.

**Review by TorstenStueber — APPROVED** [review]

File: runtime/pendulum/src/lib.rs
Comment (RustNinja) [reply runtime/pendulum/src/lib.rs:17]: @TorstenStueber 
but if you want i create a pr just few sec ago to re export only for testing
https://github.com/pendulum-chain/pendulum/pull/200

**Comment by ebma** [comment]
> for Amplitude there is no any other reserve backed assets from other chain that we should support so i decide to not touch amplitude runtime without actual requirements for it.
I see. Well, the reasoning makes sense, I would just be afraid that we miss or forget about it once we need it. But if there really is no need to have it on Amplitude yet then let's not add it there.

**Pull request merged**
